File: app/src/main/java/io/treehouses/remote/Fragments/NetworkFragment.java
package io.treehouses.remote.Fragments;

import android.annotation.SuppressLint;
import android.app.AlertDialog;
import android.graphics.Color;
import android.os.Bundle;
import android.os.Handler;
import android.os.Message;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ExpandableListView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import io.treehouses.remote.Constants;
import io.treehouses.remote.Fragments.DialogFragments.WifiDialogFragment;
import io.treehouses.remote.Network.BluetoothChatService;
import io.treehouses.remote.R;
import io.treehouses.remote.adapter.NetworkListAdapter;
import io.treehouses.remote.adapter.ViewHolderReboot;
import io.treehouses.remote.bases.BaseFragment;
import io.treehouses.remote.pojo.NetworkListItem;
import io.treehouses.remote.utils.ButtonConfiguration;

public class NetworkFragment extends BaseFragment {

    private static NetworkFragment instance = null;
    private int lastPosition = -1;
    private BluetoothChatService mChatService = null;
    private Boolean alert = true;
    private Boolean changeList = true;
    private ExpandableListView expListView;
    private NetworkListAdapter adapter;
    private ButtonConfiguration buttonConfiguration;
    View view;

    public NetworkFragment() { }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        view = inflater.inflate(R.layout.activity_network_fragment, container, false);
        expListView = view.findViewById(R.id.lvExp);
        instance = this;
        mChatService = listener.getChatService();
        mChatService.updateHandler(mHandler);
        return view;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        adapter = new NetworkListAdapter(getContext(), NetworkListItem.getNetworkList(), mChatService);
        adapter.setListener(listener);
        expListView.setOnGroupExpandListener(groupPosition -> {
            if (groupPosition == 6) {
                Log.e("TAG", "groupPosition: " + groupPosition);
                expListView.collapseGroup(6);
                updateNetworkMode();
            }

            if (lastPosition != -1 && groupPosition != lastPosition) {
                expListView.collapseGroup(lastPosition);
            }
            lastPosition = groupPosition;
        });
        expListView.setAdapter(adapter);
        expListView.setGroupIndicator(null);
    }

    @Override
    public void onResume() {
        super.onResume();
        alert = false;
        expListView.expandGroup(6);
    }

    public static NetworkFragment getInstance() {
        return instance;
    }

    public void setButtonConfiguration(ButtonConfiguration buttonConfiguration) {
        this.buttonConfiguration = buttonConfiguration;
    }

    public ButtonConfiguration getButtonConfiguration(){
        return buttonConfiguration;
    }

    private void updateNetworkMode() {
        changeList = true;
        listener.sendMessage("treehouses networkmode");
        Log.e("TAG", "network mode updated");
        Toast.makeText(getContext(), "Network Mode updated", Toast.LENGTH_SHORT).show();
    }

    public AlertDialog showAlertDialog(final String message, final String title) {
        return new AlertDialog.Builder(getContext())
                .setTitle(title)
                .setMessage(message)
                .setIcon(R.drawable.wificon)
                .setPositiveButton("No", (dialog, which) -> dialog.cancel())
                .setNegativeButton("Yes", (dialog, which) -> {
                    switch (title) {
                        case "Reset":
                            caseReset();
                            break;
                        case "Reboot":
                            ViewHolderReboot.getInstance().reboot(listener, mChatService, getContext());
                            break;
                    }
                }).show();
    }

    private void caseReset() {
        listener.sendMessage("treehouses default network");
        Thread thread = new Thread(() -> {
            try {
                Thread.sleep(2000);
                updateNetworkMode();
            } catch (Exception e) {
                e.printStackTrace();
            }
        });
        thread.start();
    }

    private AlertDialog showAlertDialog(final String message) {
        return new AlertDialog.Builder(getContext())
                .setTitle("OUTPUT:")
                .setMessage(message)
                .setIcon(R.drawable.wificon)
                .setNegativeButton("OK", (dialog, which) -> dialog.cancel()).show();
    }

    public void showWifiDialog(View v) {
        androidx.fragment.app.DialogFragment dialogFrag = WifiDialogFragment.newInstance();
        dialogFrag.setTargetFragment(this, Constants.REQUEST_DIALOG_FRAGMENT_HOTSPOT);
        dialogFrag.show(getActivity().getSupportFragmentManager().beginTransaction(), "wifiDialog");
    }

    private void changeList(String readMessage) {
        listener.sendMessage("treehouses networkmode info");
        adapter.changeList("Network Mode: " + readMessage, 6);
        switch (readMessage) {
            case "default": // ethernet
                expListView.expandGroup(0);
                adapter.changeList("Ethernet: Automatic", 0);
                break;
            case "static ethernet": // ethernet
                expListView.expandGroup(0);
                adapter.changeList("Ethernet: Manual", 0);
                break;
            case "wifi": // wifi
                expListView.expandGroup(1);
                break;
            case "ap internet": // hotspot
            case "ap local": // hotspot
                expListView.expandGroup(2);
                break;
            case "bridge": // bridge
                expListView.expandGroup(3);
                break;
        }
    }

    private Boolean btnConfigValidation(String readMessage) {
        Button btnConfig = view.findViewById(R.id.btn_start_config);
        if (buttonConfiguration == null) return false;
        switch (readMessage) {
            case "This pirateship has anchored successfully!": break; // hotspot or ethernet
            case "the bridge has been built ;), a reboot is required to apply changes": break; // bridge
            case "open wifi network": break; // wifi with no password
            case "password network": // wifi with password
                alert = false;
                buttonConfiguration.buttonProperties(true, Color.WHITE, btnConfig);
                buttonConfiguration.setMessageSent(false);
                updateNetworkMode();
                return true;
        }

        if (readMessage.contains("Error")) {
            try {
                buttonConfiguration.setMessageSent(false);
                buttonConfiguration.buttonProperties(true, Color.WHITE, btnConfig);
                alert = true;
            } catch (Exception e) {
                e.printStackTrace();
                changeList("Error");
            }
        }
        return false;
    }

    private void prefill(String readMessage) {
        if (readMessage.contains("essid")) {
            String[] array = readMessage.split(",");
            for (String element : array) {
                elementConditions(element);
            }
        }
    }

    private void elementConditions(String element) {
        Log.e("TAG", "networkmode= " + element);
        if (element.contains("wlan0") && !element.contains("ap essid")) {                   // bridge essid
            setSSIDText(element.substring(14).trim());
        } else if (element.contains("ap essid")) {                                          // ap essid
            setSSIDText(element.substring(16).trim());
        } else if (element.contains("ap0")) {                                               // hotspot essid for bridge
            ButtonConfiguration.getEtHotspotEssid().setText(element.substring(11).trim());
        } else if (element.contains("essid")) {                                             // wifi ssid
            setSSIDText(element.substring(6).trim());
        }
    }

    private void setSSIDText(String substring) {
        if (ButtonConfiguration.getSSID() != null)
            ButtonConfiguration.getSSID().setText(substring);
    }

    /**
     * The Handler that gets information back from the BluetoothChatService
     */
    @SuppressLint("HandlerLeak")
    private final Handler mHandler = new Handler() {
        @Override
        public void handleMessage(Message msg) {
            if (msg.what == Constants.MESSAGE_READ) {
                String readMessage = (String) msg.obj;
                Log.d("TAG", "readMessage = " + readMessage);

                Boolean result = btnConfigValidation(readMessage.trim());
                if (result || readMessage.trim().equals("false") || readMessage.trim().contains("true")) {
                    return;
                }

                prefill(readMessage);

                if (readMessage.contains("please reboot your device")) {
                    alert = true;
                    changeList = false;
                }

                if (changeList) {
                    changeList(readMessage.trim());
                    changeList = false;
                }

                if (alert) {
                    showAlertDialog(readMessage);
                    alert = false;
                }
            }
        }
    };

}

File: app/src/main/kotlin/io/treehouses/remote/fragments/NetworkFragment.kt
package io.treehouses.remote.fragments

import android.app.Activity
import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.os.Message
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import com.google.android.material.bottomsheet.BottomSheetDialogFragment
import io.treehouses.remote.Constants
import io.treehouses.remote.fragments.dialogfragments.bottomsheetdialogs.BridgeBottomSheet
import io.treehouses.remote.fragments.dialogfragments.bottomsheetdialogs.EthernetBottomSheet
import io.treehouses.remote.fragments.dialogfragments.bottomsheetdialogs.HotspotBottomSheet
import io.treehouses.remote.fragments.dialogfragments.bottomsheetdialogs.WifiBottomSheet
import io.treehouses.remote.fragments.dialogfragments.WifiDialogFragment
import io.treehouses.remote.interfaces.FragmentDialogInterface
import io.treehouses.remote.R
import io.treehouses.remote.Tutorials
import io.treehouses.remote.bases.BaseFragment
import io.treehouses.remote.databinding.ActivityNetworkFragmentBinding
import io.treehouses.remote.ui.home.HomeFragment
import io.treehouses.remote.utils.*

class NetworkFragment : BaseFragment(), View.OnClickListener, FragmentDialogInterface {
    private lateinit var binding: ActivityNetworkFragmentBinding
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val msg = getString(R.string.TREEHOUSES_NETWORKMODE)
        val toastMsg = "Network Mode retrieved"
        binding = ActivityNetworkFragmentBinding.inflate(inflater, container, false)
        mChatService = listener.getChatService()
        mChatService.updateHandler(mHandler)

        //update Network mode
        Utils.sendMessage(listener, Pair(msg, toastMsg), context, Toast.LENGTH_LONG)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        //Listeners
        binding.networkWifi.setOnClickListener(this)
        binding.networkHotspot.setOnClickListener(this)
        binding.networkBridge.setOnClickListener(this)
        binding.networkEthernet.setOnClickListener(this)
        binding.buttonNetworkMode.setOnClickListener(this)
        binding.rebootRaspberry.setOnClickListener(this)
        binding.resetNetwork.setOnClickListener(this)
        binding.discoverBtn.setOnClickListener(this)
        Tutorials.networkTutorials(binding, requireActivity())
    }

    private fun showBottomSheet(fragment: BottomSheetDialogFragment, tag: String) {
        fragment.setTargetFragment(this@NetworkFragment, Constants.NETWORK_BOTTOM_SHEET)
        fragment.show(requireActivity().supportFragmentManager, tag)
    }

    override fun onClick(v: View) {
        val msg = getString(R.string.TREEHOUSES_NETWORKMODE)
        val toastMsg = "Network Mode retrieved"
        when {
            binding.networkWifi == v -> showBottomSheet(WifiBottomSheet(), "wifi")
            binding.networkHotspot == v -> showBottomSheet(HotspotBottomSheet(), "hotspot")
            binding.networkBridge == v -> showBottomSheet(BridgeBottomSheet(), "bridge")
            binding.networkEthernet == v -> showBottomSheet(EthernetBottomSheet(), "ethernet")
            binding.buttonNetworkMode == v -> Utils.sendMessage(listener, Pair(msg, toastMsg), context, Toast.LENGTH_LONG)
            binding.rebootRaspberry == v -> reboot()
            binding.resetNetwork == v -> resetNetwork()
            binding.discoverBtn == v-> listener.openCallFragment(DiscoverFragment())
        }
    }

    private fun updateNetworkText(mode: String) {
        binding.currentNetworkMode.text = "Current Network Mode: $mode"
    }

    private fun performAction(output: String) {
        //Return from treehouses networkmode
        when (match(output)) {
            RESULTS.NETWORKMODE, RESULTS.DEFAULT_NETWORK -> updateNetworkText(output)

            RESULTS.DEFAULT_CONNECTED -> {

                val msg = getString(R.string.TREEHOUSES_NETWORKMODE)
                val toastMsg = "Network Mode retrieved"
                Toast.makeText(context, "Network Mode switched to default", Toast.LENGTH_LONG).show()
                //update network mode
                Utils.sendMessage(listener, Pair(msg, toastMsg), context, Toast.LENGTH_LONG)
            }
            RESULTS.ERROR -> {
                showDialog(context,"Error", output)
                binding.networkPbar.visibility = View.GONE
            }
            RESULTS.HOTSPOT_CONNECTED, RESULTS.WIFI_CONNECTED, RESULTS.BRIDGE_CONNECTED -> {
                val msg = getString(R.string.TREEHOUSES_NETWORKMODE)
                val toastMsg = "Network Mode retrieved"
                showDialog(context,"Network Switched", output)
                showDialog(context,"Network Switched", output)
                //update network mode
                Utils.sendMessage(listener, Pair(msg, toastMsg), context, Toast.LENGTH_LONG)
                binding.networkPbar.visibility = View.GONE
            }
            else -> logE("NewNetworkFragment: Result not Found")
        }
    }

    private fun rebootHelper() {
        try {
            listener.sendMessage(getString(R.string.REBOOT))
            Thread.sleep(1000)
            if (mChatService.state != Constants.STATE_CONNECTED) {
                Toast.makeText(context, "Bluetooth Disconnected: Reboot in progress", Toast.LENGTH_LONG).show()
                listener.openCallFragment(HomeFragment())
                requireActivity().title = "Home"
            } else {
                Toast.makeText(context, "Reboot Unsuccessful", Toast.LENGTH_LONG).show()
            }
        } catch (e: InterruptedException) {
            e.printStackTrace()
        }
    }

    private fun reboot() {
        val a = createAlertDialog(context, R.style.CustomAlertDialogStyle, "Reboot",
                "Are you sure you want to reboot your device?")
                .setPositiveButton("Yes") { dialog: DialogInterface, _: Int ->
                    rebootHelper()
                    dialog.dismiss()
                }.setNegativeButton("No") { dialog: DialogInterface, _: Int -> dialog.dismiss() }.create()
        a.window!!.setBackgroundDrawableResource(android.R.color.transparent)
        a.show()
    }

    private fun resetNetwork() {
        val a = createAlertDialog(context, R.style.CustomAlertDialogStyle, "Reset Network",
                "Are you sure you want to reset the network to default?")
                .setPositiveButton("Yes") { _: DialogInterface?, _: Int ->
                    val msg = getString(R.string.TREEHOUSES_DEFAULT_NETWORK)
                    val toastMsg = "Switching to default network..."
                    Utils.sendMessage(listener, Pair(msg, toastMsg), context, Toast.LENGTH_LONG)
                }.setNegativeButton("No") { dialog: DialogInterface, _: Int -> dialog.dismiss() }.create()
        a.window!!.setBackgroundDrawableResource(android.R.color.transparent)
        a.show()
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (resultCode != Activity.RESULT_OK) return
        if (requestCode == Constants.NETWORK_BOTTOM_SHEET && data!!.getBooleanExtra(CLICKED_START_CONFIG, false)) {
            binding.networkPbar.visibility = View.VISIBLE
        }
    }

    override fun getMessage(msg: Message) {
        if (msg.what == Constants.MESSAGE_READ) {
            val readMessage = msg.obj as String
            logD("readMessage = $readMessage")
            performAction(readMessage)
        }
    }

    companion object {
        @JvmField
        var CLICKED_START_CONFIG = "clicked_config"
        @JvmStatic
        fun openWifiDialog(bottomSheetDialogFragment: BottomSheetDialogFragment, context: Context?) {
            if (Build.VERSION.SDK_INT <= Build.VERSION_CODES.LOLLIPOP_MR1) {
                Toast.makeText(context, "Wifi scan requires at least android API 23", Toast.LENGTH_LONG).show()
            } else {
                val dialogFrag = WifiDialogFragment.newInstance()
                dialogFrag.setTargetFragment(bottomSheetDialogFragment, Constants.REQUEST_DIALOG_WIFI)
                dialogFrag.show(bottomSheetDialogFragment.requireActivity().supportFragmentManager.beginTransaction(), "wifiDialog")
            }
        } //    Next Version:
        //    private void elementConditions(String element) {
        //        Log.e("TAG", "networkmode= " + element);
        //        if (element.contains("wlan0") && !element.contains("ap essid")) {                   // bridge essid
        //            setSSIDText(element.substring(14).trim());
        //        } else if (element.contains("ap essid")) {                                          // ap essid
        //            setSSIDText(element.substring(16).trim());
        //        } else if (element.contains("ap0")) {                                               // hotspot essid for bridge
        //            ButtonConfiguration.getEtHotspotEssid().setText(element.substring(11).trim());
        //        } else if (element.contains("essid")) {                                             // wifi ssid
        //            setSSIDText(element.substring(6).trim());
        //        }
        //    }
    }
}